newsrevenuehub/rev-engine
# DEV-1021 enable feature flagging

Please fill out each section even if it's just with "N/A"

#### Plan? (if this draft/incomplete indicate what you intend to do and how)

This PR is the first in several that will bring feature flagging to the front- and back-end for rev-engine.

This PR adds django-waffle and integrates feature flagging with the permissions layer (initially for a single endpoint: contributions). 

Later PRs will expose a custom flag model created in this PR as an API resource which can be used in the SPA to gate access to features there. 


#### What's this PR do?

- Adds a custom `Flag` model (which initially is functional equivalent of the default waffle flag) that can later be extended to integrate with roleassignment (so flags can be rolled out for entire programs, etc.).
- Adds an admin model for `Flag`
- Adds a `HasFlaggedAccessToContributionsApiResource` permission that is used to gate the contributions api endpoint to users for whom the related flag is not active. 
- Uses the new feature flag approach to gate access to the contributions API resource (not inclusive of stripe specific API endpoints).
- Adds two tests to `apps.contributions.tests.test_views.TestContributionsViewSet` around feature flagging. The first is the "happy path", and the second shows behavior if `HasFlaggedAccessToContributionsApiResource` can't find its required flag on initialization.


#### Why are we doing this? How does it help us?

Described in detail on JIRA.


#### How should this be manually tested?

(first three steps are only if you're running locally)

1. Check out this branch.
2. Run migrations
3. Run dev server 
4. In the Django admin, go to the Flag admin, and create a flag whose name is the same value as `apps.flags.constants.CONTRIBUTIONS_API_ENDPOINT_ACCESS_FLAG_NAME` is set to (in source code right now, that's "contributions-api-endpoint-access"). Set `.everyone` to False, `superusers` to True, and add the user ID for a non superuser to `.users`. Save.
5. Try logging in as a superuser, then go to `/api/v1/contributions/` in a browser tab. You should be able to access. Alternatively, you can log in to the SPA as a superuser, and the contributions tab should populate with any contributions that are in the db.
6. Try logging in as a non super user who hasn't been explicitly added. Follow the same steps as 5. This time the API should return a 403 error.
7. Try logging in as the explicitly added user from step 4. Follow same steps as 5. This time the API should return any results (and 200).
8. In the Django admin, update the feature flag's `everyone` value to True. 
9. Repeat step 6. This time the API should 200 and return any results that user is allowed to see.

#### Have automated unit tests been added?

Yes

#### How should this change be communicated to end users?

#### Are there any smells or added technical debt to note?

#### Has this been documented? If so, where?

I don't believe so. Once this and initial, related SPA-side feature flag tickets are complete, would make sense to consolidate knowledge around feature flagging in Confluence.

#### What are the relevant tickets?
**(Note: Please use syntax [JIRA-123] to link any relevant tickets)**

[DEV-1021]

#### Do any changes need to be made before deployment to production (adding environment variables, for example)?


#### Are there next steps? If so, what? Have tickets been opened for them?

After this is merged, we'll need to create a feature flag with name set to the value of `apps.flags.constants.CONTRIBUTIONS_API_ENDPOINT_ACCESS_FLAG_NAME`. That flag should have `superusers` set to True, `everyone` to "Unknown", and any individual users should have their IDs added to the `users` field in the admin.
